QuadMeshRectilinear returning all NaN in some cases

<!--
Thanks for contacting us! Please read and follow these instructions carefully, then you can delete this introductory text. Note that the issue tracker is NOT the place for usage questions and technical assistance; post those at [Discourse](https://discourse.holoviz.org) instead. Issues without the required information below may be closed immediately.
-->


This one is pretty weird :/, either broadcasting out coordinates or sorting the y-coordinate makes it work.

#### ALL software version info
(this library, plus any other relevant software, e.g. bokeh, python, notebook, OS, browser, etc should be added within the dropdown below.)

<details>
    <summary>Software Version Info</summary>

```plaintext
datashader 0.18.2
macos
numba 0.61.2
```
</details>

#### Description of expected behavior and the observed behavior

#### Complete, minimal, self-contained example code that reproduces the issue

```python
import datashader as dsh
import numpy as np
import xarray as xr

west = 3125000.0
south = 3250000.0
east = 4250000.0
north = 4375000.0

da = xr.DataArray(
    np.random.random((940, 940)),
    dims=("x", "y"),
    coords={
        "x": np.linspace(3123580.0, 4250380.0, 940),
        "y": np.linspace(4376200.0, 3249400.0, 940),
    },
    name="foo",
)
cvs = dsh.Canvas(256, 256, x_range=(west, east), y_range=(south, north))
# uncommenting this line makes it work
#da = da.assign_coords(dict(zip(("x", "y"), xr.broadcast(da["x"], da["y"]))))
# uncommenting this line makes it work, but with different results
#da = da.isel(y=slice(None, None, -1))
cvs.quadmesh(da.transpose("y", "x"), x="x", y="y")

```

#### Stack traceback and/or browser JavaScript console output

#### Screenshots or screencasts of the bug in action

- [ ] I may be interested in making a pull request to address this

Both this and #1438 seem to happen because of the swap of bounds here:

https://github.com/holoviz/datashader/blob/73f686c34173419aadd2135a97b78496b55f17be/datashader/glyphs/quadmesh.py#L86-L88

Not entirely sure why... And if this is the cause of the problem. Will mark both as a bug for now. 


Commenting out the lines and running the test suite will cause the following two tests to fail:
```
FAILED datashader/tests/test_quadmesh.py::test_raster_quadmesh_autorange_reversed[numpy] - AssertionError: Left and right DataArray objects are not equal
FAILED datashader/tests/test_quadmesh.py::test_raster_quadmesh_autorange_reversed[dask.array] - AssertionError: Left and right DataArray objects are not equal
```

--- a/datashader/glyphs/quadmesh.py
+++ b/datashader/glyphs/quadmesh.py
@@ -308,26 +308,30 @@ class QuadMeshRaster(QuadMeshRectilinear):
                 offset_x, offset_y, out_w, out_h, *aggs_and_cols
         ):
             for out_j in prange(out_h):
-                src_j0 = int(max(
-                    math.floor(scale_y * (out_j + 0.0) + translate_y - offset_y), 0
-                ))
-                src_j1 = int(min(
-                    math.floor(scale_y * (out_j + 1.0) + translate_y - offset_y), src_h
-                ))
-                # Handle negative scale_y (descending coordinates)
-                if scale_y < 0 and src_j0 > src_j1:
-                    src_j0, src_j1 = src_j1, src_j0
+                # Calculate raw indices first
+                raw_j0 = math.floor(scale_y * (out_j + 0.0) + translate_y - offset_y)
+                raw_j1 = math.floor(scale_y * (out_j + 1.0) + translate_y - offset_y)
+
+                # Handle negative scale_y (descending coordinates) - swap before clamping
+                if scale_y < 0 and raw_j0 > raw_j1:
+                    raw_j0, raw_j1 = raw_j1, raw_j0
+
+                # Now clamp to valid range
+                src_j0 = int(max(raw_j0, 0))
+                src_j1 = int(min(raw_j1, src_h))
 
                 for out_i in range(out_w):
-                    src_i0 = int(max(
-                        math.floor(scale_x * (out_i + 0.0) + translate_x - offset_x), 0
-                    ))
-                    src_i1 = int(min(
-                        math.floor(scale_x * (out_i + 1.0) + translate_x - offset_x), src_w
-                    ))
-                    # Handle negative scale_x (descending coordinates)
-                    if scale_x < 0 and src_i0 > src_i1:
-                        src_i0, src_i1 = src_i1, src_i0
+                    # Calculate raw indices first
+                    raw_i0 = math.floor(scale_x * (out_i + 0.0) + translate_x - offset_x)
+                    raw_i1 = math.floor(scale_x * (out_i + 1.0) + translate_x - offset_x)
+
+                    # Handle negative scale_x (descending coordinates) - swap before clamping
+                    if scale_x < 0 and raw_i0 > raw_i1:
+                        raw_i0, raw_i1 = raw_i1, raw_i0
+
+                    # Now clamp to valid range
+                    src_i0 = int(max(raw_i0, 0))
+                    src_i1 = int(min(raw_i1, src_w))
 
                     for src_j in range(src_j0, src_j1):
                         for src_i in range(src_i0, src_i1):
@@ -341,25 +345,29 @@ class QuadMeshRaster(QuadMeshRectilinear):
         ):
             out_i, out_j = cuda.grid(2)
             if out_i < out_w and out_j < out_h:
-                src_j0 = max(
-                    math.floor(scale_y * (out_j + 0.0) + translate_y - offset_y), 0
-                )
-                src_j1 = min(
-                    math.floor(scale_y * (out_j + 1.0) + translate_y - offset_y), src_h
-                )
-                # Handle negative scale_y (descending coordinates)
-                if scale_y < 0 and src_j0 > src_j1:
-                    src_j0, src_j1 = src_j1, src_j0
+                # Calculate raw indices first
+                raw_j0 = math.floor(scale_y * (out_j + 0.0) + translate_y - offset_y)
+                raw_j1 = math.floor(scale_y * (out_j + 1.0) + translate_y - offset_y)
 
-                src_i0 = max(
-                    math.floor(scale_x * (out_i + 0.0) + translate_x - offset_x), 0
-                )
-                src_i1 = min(
-                    math.floor(scale_x * (out_i + 1.0) + translate_x - offset_x), src_w
-                )
-                # Handle negative scale_x (descending coordinates)
-                if scale_x < 0 and src_i0 > src_i1:
-                    src_i0, src_i1 = src_i1, src_i0
+                # Handle negative scale_y (descending coordinates) - swap before clamping
+                if scale_y < 0 and raw_j0 > raw_j1:
+                    raw_j0, raw_j1 = raw_j1, raw_j0
+
+                # Now clamp to valid range
+                src_j0 = max(raw_j0, 0)
+                src_j1 = min(raw_j1, src_h)
+
+                # Calculate raw indices first
+                raw_i0 = math.floor(scale_x * (out_i + 0.0) + translate_x - offset_x)
+                raw_i1 = math.floor(scale_x * (out_i + 1.0) + translate_x - offset_x)
+
+                # Handle negative scale_x (descending coordinates) - swap before clamping
+                if scale_x < 0 and raw_i0 > raw_i1:
+                    raw_i0, raw_i1 = raw_i1, raw_i0
+
+                # Now clamp to valid range
+                src_i0 = max(raw_i0, 0)
+                src_i1 = min(raw_i1, src_w)
 
                 for src_j in range(src_j0, src_j1):
                     for src_i in range(src_i0, src_i1):

--- a/datashader/glyphs/quadmesh.py
+++ b/datashader/glyphs/quadmesh.py
@@ -321,18 +321,12 @@ class QuadMeshRaster(QuadMeshRectilinear):
                 src_j1 = int(min(raw_j1, src_h))
 
                 for out_i in range(out_w):
-                    # Calculate raw indices first
-                    raw_i0 = math.floor(scale_x * (out_i + 0.0) + translate_x - offset_x)
-                    raw_i1 = math.floor(scale_x * (out_i + 1.0) + translate_x - offset_x)
-
-                    # Handle negative scale_x (descending coordinates) - swap before clamping
-                    if scale_x < 0 and raw_i0 > raw_i1:
-                        raw_i0, raw_i1 = raw_i1, raw_i0
-
-                    # Now clamp to valid range
-                    src_i0 = int(max(raw_i0, 0))
-                    src_i1 = int(min(raw_i1, src_w))
-
+                    src_i0 = int(max(
+                        math.floor(scale_x * (out_i + 0.0) + translate_x - offset_x), 0
+                    ))
+                    src_i1 = int(min(
+                        math.floor(scale_x * (out_i + 1.0) + translate_x - offset_x), src_w
+                    ))
                     for src_j in range(src_j0, src_j1):
                         for src_i in range(src_i0, src_i1):
                             append(src_j, src_i, out_i, out_j, *aggs_and_cols)
@@ -357,18 +351,12 @@ class QuadMeshRaster(QuadMeshRectilinear):
                 src_j0 = max(raw_j0, 0)
                 src_j1 = min(raw_j1, src_h)
 
-                # Calculate raw indices first
-                raw_i0 = math.floor(scale_x * (out_i + 0.0) + translate_x - offset_x)
-                raw_i1 = math.floor(scale_x * (out_i + 1.0) + translate_x - offset_x)
-
-                # Handle negative scale_x (descending coordinates) - swap before clamping
-                if scale_x < 0 and raw_i0 > raw_i1:
-                    raw_i0, raw_i1 = raw_i1, raw_i0
-
-                # Now clamp to valid range
-                src_i0 = max(raw_i0, 0)
-                src_i1 = min(raw_i1, src_w)
-
+                src_i0 = max(
+                    math.floor(scale_x * (out_i + 0.0) + translate_x - offset_x), 0
+                )
+                src_i1 = min(
+                    math.floor(scale_x * (out_i + 1.0) + translate_x - offset_x), src_w
+                )
                 for src_j in range(src_j0, src_j1):
                     for src_i in range(src_i0, src_i1):
                         append(src_j, src_i, out_i, out_j, *aggs_and_cols)

--- a/datashader/glyphs/quadmesh.py
+++ b/datashader/glyphs/quadmesh.py
@@ -314,6 +314,10 @@ class QuadMeshRaster(QuadMeshRectilinear):
                 src_j1 = int(min(
                     math.floor(scale_y * (out_j + 1.0) + translate_y - offset_y), src_h
                 ))
+                # Handle negative scale_y (descending coordinates)
+                if scale_y < 0 and src_j0 > src_j1:
+                    src_j0, src_j1 = src_j1, src_j0
+
                 for out_i in range(out_w):
                     src_i0 = int(max(
                         math.floor(scale_x * (out_i + 0.0) + translate_x - offset_x), 0
@@ -321,6 +325,10 @@ class QuadMeshRaster(QuadMeshRectilinear):
                     src_i1 = int(min(
                         math.floor(scale_x * (out_i + 1.0) + translate_x - offset_x), src_w
                     ))
+                    # Handle negative scale_x (descending coordinates)
+                    if scale_x < 0 and src_i0 > src_i1:
+                        src_i0, src_i1 = src_i1, src_i0
+
                     for src_j in range(src_j0, src_j1):
                         for src_i in range(src_i0, src_i1):
                             append(src_j, src_i, out_i, out_j, *aggs_and_cols)
@@ -339,12 +347,20 @@ class QuadMeshRaster(QuadMeshRectilinear):
                 src_j1 = min(
                     math.floor(scale_y * (out_j + 1.0) + translate_y - offset_y), src_h
                 )
+                # Handle negative scale_y (descending coordinates)
+                if scale_y < 0 and src_j0 > src_j1:
+                    src_j0, src_j1 = src_j1, src_j0
+
                 src_i0 = max(
                     math.floor(scale_x * (out_i + 0.0) + translate_x - offset_x), 0
                 )
                 src_i1 = min(
                     math.floor(scale_x * (out_i + 1.0) + translate_x - offset_x), src_w
                 )
+                # Handle negative scale_x (descending coordinates)
+                if scale_x < 0 and src_i0 > src_i1:
+                    src_i0, src_i1 = src_i1, src_i0
+
                 for src_j in range(src_j0, src_j1):
                     for src_i in range(src_i0, src_i1):
                         append(src_j, src_i, out_i, out_j, *aggs_and_cols)

--- a/datashader/tests/test_quadmesh.py
+++ b/datashader/tests/test_quadmesh.py
@@ -30,7 +30,7 @@ except ImportError:
 try:
     import cudf
     import cupy
-    array_modules.append(cupy)
+    array_modules.append(pytest.param(cupy, marks=pytest.mark.gpu))
 except ImportError:
     cudf = None
     cupy = None
@@ -916,4 +916,4 @@ def test_raster_quadmesh_descending_coords_2(array_module):
     cvs = ds.Canvas(256, 256, x_range=(west, east), y_range=(south, north))
     actual = cvs.quadmesh(da.transpose("y", "x"), x="x", y="y")
     expected = cvs.quadmesh(da.isel(y=slice(None, None, -1)).transpose("y", "x"), x="x", y="y")
-    np.testing.assert_allclose(expected, actual)
+    assert_eq_xr(expected, actual, close=True)

--- a/datashader/tests/test_quadmesh.py
+++ b/datashader/tests/test_quadmesh.py
@@ -890,3 +890,30 @@ def test_raster_quadmesh_descending_coords(array_module):
 
     result = cvs.quadmesh(da.isel(y=slice(None, None, -1)).transpose("y", "x"), x="x", y="y")
     assert result.isnull().sum().item() == 0
+
+
+@pytest.mark.parametrize('array_module', array_modules)
+def test_raster_quadmesh_descending_coords_2(array_module):
+    """
+    Regression test for https://github.com/holoviz/datashader/issues/1439
+    """
+    west=3125000.0
+    south=4375000.0
+    east=4250000.0
+    north=5500000.0
+
+    # Create data with descending y coordinates (high to low)
+    da = xr.DataArray(
+        array_module.ones((940, 868)),
+        dims=("x", "y"),
+        coords={
+            "x": np.linspace(3123580.0, 4250380.0, 940),
+            "y": np.linspace(5415400.0, 4375000.0, 868),  # descending!
+        },
+        name="foo",
+    )
+
+    cvs = ds.Canvas(256, 256, x_range=(west, east), y_range=(south, north))
+    actual = cvs.quadmesh(da.transpose("y", "x"), x="x", y="y")
+    expected = cvs.quadmesh(da.isel(y=slice(None, None, -1)).transpose("y", "x"), x="x", y="y")
+    np.testing.assert_allclose(expected, actual)

--- a/datashader/tests/test_quadmesh.py
+++ b/datashader/tests/test_quadmesh.py
@@ -861,3 +861,32 @@ def test_infer_interval_breaks_2d_consistency(spacings, start_value):
 
     # Compare results
     np.testing.assert_allclose(actual, expected, rtol=1e-10, atol=1e-10)
+
+
+@pytest.mark.parametrize('array_module', array_modules)
+def test_raster_quadmesh_descending_coords(array_module):
+    """
+    Regression test for https://github.com/holoviz/datashader/issues/1439
+    """
+    west = 3125000.0
+    south = 3250000.0
+    east = 4250000.0
+    north = 4375000.0
+
+    # Create data with descending y coordinates (high to low)
+    da = xr.DataArray(
+        array_module.ones((940, 940)),
+        dims=("x", "y"),
+        coords={
+            "x": np.linspace(3123580.0, 4250380.0, 940),
+            "y": np.linspace(4376200.0, 3249400.0, 940),  # descending!
+        },
+        name="foo",
+    )
+
+    cvs = ds.Canvas(256, 256, x_range=(west, east), y_range=(south, north))
+    result = cvs.quadmesh(da.transpose("y", "x"), x="x", y="y")
+    assert result.isnull().sum().item() == 0
+
+    result = cvs.quadmesh(da.isel(y=slice(None, None, -1)).transpose("y", "x"), x="x", y="y")
+    assert result.isnull().sum().item() == 0

